### PR TITLE
Allow torch.hub.load() to load models from any local directory with a hubconf.py

### LINKED ISSUE #43622: Add a torch.hub.load_local() function that can load models from any local directory with a hubconf.py
## 🚀 Feature
A `torch.hub.load_local()` function that can load models from any local directory with a `hubconf.py`.

## Motivation
`torch.hub.load()` currently allows loading model definitions from a GitHub repo that has a `hubconf.py` file. It does this by downloading the repo, and then loading functions from `hubconf.py`.

Since it only accepts locations of the form `<repo_owner>/<repo_name>`, it is limited to repos hosted on Github.

This means it does not support the following functionality/use cases:
- The ability to use `torch.hub` on machines that do not have internet access.
- The ability to archive model definition code locally.
- The ability to load models from local directories without having to push them to Github, especially during development when you want to quickly test out the changes you make to the model definition code.

## Pitch
It will cover the above use cases.

`torch.hub.load_local()` should be trivial to implement. It will have the same signature as `torch.hub.load()`, with the first argument, `github`, replaced by `repo_dir` (which will be a local path). In terms of code, the only difference from `torch.hub.load()` will be the removal of the line that downloads the GitHub repo (https://github.com/pytorch/pytorch/blob/master/torch/hub.py#L343) and the associated `force_reload` argument (https://github.com/pytorch/pytorch/blob/master/torch/hub.py#L338).

I would probably be implementing this functionality for [azavea/raster-vision#975](https://github.com/azavea/raster-vision/issues/975) and would be happy to open a PR here, if there is interest.

## Alternatives

A more general, though harder to implement, solution would be a `torch.hub.load_uri()` function that can fetch code from any URI, local or remote.

cc @ailzhang

### PR DESCRIPTION
**Pull request by AdeelH**
Fixes #43622 

- Moves the model loading part of `torch.hub.load()` into a new ~`torch.hub._load_local()`~ `torch.hub._load_local()` function that takes in a path to a local directory that contains a `hubconf.py` instead of a repo name. 
- ~Refactors `torch.hub.load()` so that it now calls `torch.hub.load_local()` after downloading and extracting the repo.~
- Refactors `torch.hub.load()` so that it can now take in a local path instead of a repo name. Adds a `source='github'|'local'` param to choose between the two behaviors.
- Updates `torch.hub` docs to include the new function + minor fixes.

### CODE REVIEW

**Comment by AdeelH** [comment]
@ailzhang, here's the PR. I don't believe the CI errors are related.

**Review by ailzhang — COMMENTED** [review]
Thanks! Overall looks good! 
Btw I feel having something like `torch.load(<path_to_local_dir>|<github_repo>, source='github'|'local')` could be a better unified user facing API? I'm definitely open to suggestions, let me know!

**Comment by AdeelH** [comment]
>Btw I feel having something like torch.load(<path_to_local_dir>|<github_repo>, source='github'|'local') could be a better unified user facing API? I'm definitely open to suggestions, let me know!

I don't have a strong opinion either way. The `source=` way might be easier to extend if we want to add more sources in the future (say, a `gitlab` or a general `uri` source) without cluttering the API.

**Comment by AdeelH** [comment]
I have changed it to the `source='github'|'local'` approach and I agree that it does look better from an API perspective.

I have also taken the liberty to expand out the documentation for `load()` a little bit. Let me know if you agree with the changes.
image

File: torch/hub.py [hunk 1]
```diff
@@ -312,45 +312,90 @@ def help(github, model, force_reload=False):
...
-            Example: 'pytorch/vision[:hub]'
-        model (string): a string of entrypoint name defined in repo's hubconf.py
-        *args (optional): the corresponding args for callable `model`.
-        force_reload (bool, optional): whether to force a fresh download of github repo unconditionally.
-            Default is `False`.
-        verbose (bool, optional): If False, mute messages about hitting local caches. Note that the message
-            about first download is cannot be muted.
-            Default is `True`.
-        **kwargs (optional): the corresponding kwargs for callable `model`.
+        repo_or_dir (string): repo name (``repo_owner/repo_name[:tag_name]``)
```
Comment (ailzhang) [inline torch/hub.py:330]: It'll be great to clarify `repo_owner/repo_name[:tag_name]` is for `source=github` and `path` is for `source=local`.

**Review by ailzhang — APPROVED** [review]
Looks great! 
I think this is good to go after adding a test for `hub.load(source='local')` API in https://github.com/pytorch/pytorch/blob/master/test/test_utils.py#L585.

**Pull request closed without merging**